funsor/tensor.py

```diff
@@ -0,0 +1,973 @@
+    return functools.partial(_function, inputs, output)
+
+
+class Einsum(Funsor):
+    """
+    Wrapper around :func:`torch.einsum` to operate on real-valued Funsors.
+
+    Note this operates only on the ``output`` tensor. To perform sum-product
+    contractions on named dimensions, instead use ``+`` and
+    :class:`~funsor.terms.Reduce`.
+
+    :param str equation: An :func:`torch.einsum` equation.
+    :param tuple operands: A tuple of input funsors.
+    """
+    def __init__(self, equation, operands):
+        assert isinstance(equation, str)
+        assert isinstance(operands, tuple)
+        assert all(isinstance(x, Funsor) for x in operands)
+        ein_inputs, ein_output = equation.split('->')
+        ein_inputs = ein_inputs.split(',')
+        size_dict = {}
+        inputs = OrderedDict()
+        assert len(ein_inputs) == len(operands)
+        for ein_input, x in zip(ein_inputs, operands):
+            assert x.dtype == 'real'
+            inputs.update(x.inputs)
+            assert len(ein_input) == len(x.output.shape)
+            for name, size in zip(ein_input, x.output.shape):
+                other_size = size_dict.setdefault(name, size)
+                if other_size != size:
+                    raise ValueError("Size mismatch at {}: {} vs {}"
+                                     .format(name, size, other_size))
+        output = reals(*(size_dict[d] for d in ein_output))
+        super(Einsum, self).__init__(inputs, output)
+        self.equation = equation
+        self.operands = operands
+
+    def __repr__(self):
+        return 'Einsum({}, {})'.format(repr(self.equation), repr(self.operands))
+
+    def __str__(self):
+        return 'Einsum({}, {})'.format(repr(self.equation), str(self.operands))
+
+
+@eager.register(Einsum, str, tuple)
+def eager_einsum(equation, operands):
+    if all(isinstance(x, Tensor) for x in operands):
+        # Make new symbols for inputs of operands.
+        inputs = OrderedDict()
+        for x in operands:
+            inputs.update(x.inputs)
+        symbols = set(equation)
+        get_symbol = iter(map(opt_einsum.get_symbol, itertools.count()))
+        new_symbols = {}
+        for k in inputs:
+            symbol = next(get_symbol)
+            while symbol in symbols:
+                symbol = next(get_symbol)
+            symbols.add(symbol)
+            new_symbols[k] = symbol
+
+        # Manually broadcast using einsum symbols.
+        assert '.' not in equation
+        ins, out = equation.split('->')
+        ins = ins.split(',')
+        ins = [''.join(new_symbols[k] for k in x.inputs) + x_out
+               for x, x_out in zip(operands, ins)]
+        out = ''.join(new_symbols[k] for k in inputs) + out
+        equation = ','.join(ins) + '->' + out
+
+        data = ops.einsum(equation, *[x.data for x in operands])
+        return Tensor(data, inputs)
+
+    return None  # defer to default implementation
+
+
+def numeric_tensordot(x, y, dims):
```

Comment:
`torch_tensordot` is renamed to `numeric_tensordot`. Alternatives are welcomed!

Reply:
I think simply `tensordot()` should work now? The previous `torch_` prefix was simply to distinguish it from `numpy_tensordot()`.